# DATA CLEANING

In [1]:
#immporting necessary libraries
import pandas as pd
import numpy as np
from IPython.display import display

pd.options.display.max_columns = None
pd.options.display.max_rows = None



In [2]:
data = pd.read_csv('listings.csv',low_memory=False)

In [3]:
data.shape

(21235, 106)

There are 21,235 samples and 106 features in the dataset

Let's look into the dataset now

In [4]:
data.head(3)

id                         listing_url       scrape_id last_scraped  \
0   2078   https://www.airbnb.com/rooms/2078  20190813195408   2019-08-14   
1   2843   https://www.airbnb.com/rooms/2843  20190813195408   2019-08-14   
2  14584  https://www.airbnb.com/rooms/14584  20190813195408   2019-08-14   

                                                name  \
0                Central comfy close to Metro Laurie   
1                     Fabulous downtown/airport room   
2  Spread Out in 1,200 Square Feet of Luxury in t...   

                                             summary  \
0  Centrally located in trendy Le Plateau.   Join...   
1  Come stay in this trendy area of downtown Mont...   
2  This stylish and spacious loft has modern artw...   

                                               space  \
0  Centrally located  clean comfy furnished room ...   
1  Private individual room in the downtown area, ...   
2  It's artistic and tasteful with a modern touch...   

                                         description experiences_offered  \
0  Centrally located in trendy Le Plateau.   Join...                none   
1  Come stay in this trendy area of downtown Mont...                none   
2  This stylish and spacious loft has modern artw...                none   

                               neighborhood_overview  \
0  Although my place is centrally located in Le P...   
1  This area of St Henri is actually downtown Mon...   
2  This chic apartment can be found in the Platea...   

                                               notes  \
0  I always inform my guests the places/attractio...   
1  From this room you can walk to downtown Montre...   
2                                                NaN   

                                             transit  \
0  I am 6-8 mins. walk to Metro Laurie and bus ro...   
1  There is a major metro/subway stop called Lion...   
2                                                NaN   

                                              access  \
0  You will have full access in the apartment.  I...   
1  The guests can use the kitchen anytime, fully ...   
2                                                NaN   

                                         interaction  \
0  I am always available to help and give informa...   
1  I meet guests when they arrive show them the h...   
2                                                NaN   

                                         house_rules  thumbnail_url  \
0  No house rules but I expect my guests to follo...            NaN   
1  No smoking and no pets Airport bus stops acros...            NaN   
2                                                NaN            NaN   

   medium_url                                        picture_url  \
0         NaN  https://a0.muscache.com/im/pictures/88170231/0...   
1         NaN  https://a0.muscache.com/im/pictures/89b0268d-2...   
2         NaN  https://a0.muscache.com/4ea/air/v2//pictures/0...   

   xl_picture_url  host_id                                 host_url host_name  \
0             NaN     2277   https://www.airbnb.com/users/show/2277     Nelia   
1             NaN     2319   https://www.airbnb.com/users/show/2319      Gail   
2             NaN    57358  https://www.airbnb.com/users/show/57358    Verona   

   host_since             host_location  \
0  2008-08-19  Montreal, Québec, Canada   
1  2008-08-19  Montreal, Québec, Canada   
2  2009-11-25  Montreal, Québec, Canada   

                                          host_about  host_response_time  \
0  Honest, quiet and adhere to the golden rules a...                 NaN   
1  Manage a network of downtown rooms and private...      within an hour   
2  I am a half Filipino/Norwegian painter living ...  within a few hours   

  host_response_rate  host_acceptance_rate host_is_superhost  \
0                NaN                   NaN                 f   
1               100%                   NaN                 f   
2               100%                   

The types for each data feature

In [5]:
print(data.dtypes)

id                                                int64
listing_url                                      object
scrape_id                                         int64
last_scraped                                     object
name                                             object
summary                                          object
space                                            object
description                                      object
experiences_offered                              object
neighborhood_overview                            object
notes                                            object
transit                                          object
access                                           object
interaction                                      object
house_rules                                      object
thumbnail_url                                   float64
medium_url                                      float64
picture_url                                     

We see here that data types for price and cleaning fee and security_deposit are not right, they are supposed to be numbers but they are object here.



In [6]:
data['price'] = data['price'].str.replace(',', '')
data['price'] = data['price'].str.replace('$', '')
data['price'] = data['price'].astype(float)


data['security_deposit'] = data['security_deposit'].str.replace(',', '')
data['security_deposit'] = data['security_deposit'].str.replace('$', '')
data['security_deposit'] = data['security_deposit'].astype(float)

data['cleaning_fee'] = data['cleaning_fee'].str.replace(',', '')
data['cleaning_fee'] = data['cleaning_fee'].str.replace('$', '')
data['cleaning_fee'] = data['cleaning_fee'].astype(float)



Checking which features have null values and how many?

In [7]:
data.isnull().sum()

id                                                  0
listing_url                                         0
scrape_id                                           0
last_scraped                                        0
name                                               18
summary                                           942
space                                            7008
description                                       511
experiences_offered                                 0
neighborhood_overview                            8187
notes                                           13646
transit                                          8101
access                                          10576
interaction                                      9686
house_rules                                      9099
thumbnail_url                                   21235
medium_url                                      21235
picture_url                                         0
xl_picture_url              

Many columns like thumbnail_url, medium_url, xl_picture_url, host_acceptance_rate, neighbourhood_group_cleansed, license have all null values. 

Moreover columns like  weekly_price, monthly_price, and square_feet are almost filled with null values

We can drop the above mentioned columns

There are many other columns that have significant number of null values such as security_deposit, cleaning_fee, host_about, host_response_time, host_response_rate, space,neighborhood_overview, transit,access, notes, interaction, house_rules. Such columns need further inspection whether they are important or not , whether they can be imputed or not, can be decided later


In [8]:
drop_columns=['thumbnail_url', 'medium_url', 'xl_picture_url', 'host_acceptance_rate', 'neighbourhood_group_cleansed', 'license', 'square_feet','weekly_price','monthly_price']

data = data.drop(drop_columns,axis=1)

Lets dig deeper and look into the values the important features hold

Starting with the country feature

In [9]:
data.country.value_counts()

Canada    21232
France        3
Name: country, dtype: int64

In [10]:
data[data['country']=='France']

id                            listing_url       scrape_id  \
16032  31391340  https://www.airbnb.com/rooms/31391340  20190813195408   
16082  31508450  https://www.airbnb.com/rooms/31508450  20190813195408   
16573  32566938  https://www.airbnb.com/rooms/32566938  20190813195408   

      last_scraped                                               name  \
16032   2019-08-13              Chambre Plateau St-Laurent / St-Denis   
16082   2019-08-14  Beautiful room with balcony in Plateau-Mont-Royal   
16573   2019-08-13                            Chambre rue Saint Denis   

                                                 summary  \
16032  Découvrez le charme de cet appartement , situé...   
16082  This apartment with room have two balcony. is ...   
16573  Chambre privée dans une grande colocation étud...   

                                                   space  \
16032                    Cuisine - Salon - Salle de bain   
16082  very bright apartment fully equipped with kitc...   
16573   Grand logement trés bien placé, proche de tout !   

                                             description experiences_offered  \
16032  Découvrez le charme de cet appartement , situé...                none   
16082  This apartment with room have two balcony. is ...                none   
16573  Chambre privée dans une grande colocation étud...                none   

                                   neighborhood_overview  \
16032  Plateau-Mont-Royal / Down-Town  est un quartie...   
16082                                                NaN   
16573  Principal quartier culturel de la ville, le Qu...   

                                                   notes  \
16032  Il y a un chat nommé Nala qui se promène dans ...   
16082                                                NaN   
16573                                                NaN   

                                                 transit  \
16032  4 mins du Metro Sherbrooke ou Mont-Royal Ligne...   
16082                                                NaN   
16573  Bus et Métro à proximité  Ligne Verte, Jaune e...   

                                                  access  \
16032                   Cuisine / Salon / Salle de bain.   
16082  you have access to all the rooms thank you for...   
16573  Accés à la chambre, la cuisine, la salle de ba...   

                                             interaction  \
16032  Je me tiens à votre dispositions pour toutes q...   
16082  my friend Edouard will welcome you to the apar...   
16573  Mes colocs sont très sympas et vous accueiller...   

                                             house_rules  \
16032                                                NaN   
16082  you can use everything that is available in th...   
16573                                                NaN   

                                             picture_url    host_id  \
16032  https://a0.muscache.com/im/pictures/9399cf8e-0...   11685689   
16082  https://a0.muscache.com/im/pictures/1f3ec26a-a...   72878131   
16573  https://a0.muscache.com/im/pictures/bc3334e4-f...  241717209   

                                          host_url host_name  host_since  \
16032   https://www.airbnb.com/users/show/11685689    Thomas  2014-01-27   
16082   https://www.airbnb.com/users/show/72878131   Yannick  2016-05-19   
16573  https://www.airbnb.com/users/show/241717209   Sixtine  2019-02-07   

                                  host_location  \
16032  Nice, Provence-Alpes-Côte d'Azur, France   
16082                  Montreal, Quebec, Canada   
16573         Versailles, Île-de-France, France   

                                              host_about host_response_time  \
16032                                                NaN                NaN   
16082                                                NaN       within a day   
16573  Étudiante en design graphique, je voyage assez...                NaN   

      host_response_rate host_is_superhost  \
16032           

We can see there are 3 instances that have their country set as France, if we further look into the data for those rows, we find that the listing belongs to Canada only. Since there were just 3 such case we could manually look up and verify, if there were more such noisy data, best option would be to drop such columns as we have enough data.

In [11]:
# if you want to delete rows with France as country
#data = data[data.country == 'Canada']

#change France to Canada
data.country.replace('France','Canada',inplace=True)

#verifying the modification
data.country.value_counts()

Canada    21235
Name: country, dtype: int64

Looking into the country code

In [12]:
data.country_code.value_counts()

CA    21232
FR        3
Name: country_code, dtype: int64

Applying the same procedure as we applied to the city column

In [13]:
#change France to Canada
data.country_code.replace('FR','CA',inplace=True)

#verifying the modification
data.country_code.value_counts()

CA    21235
Name: country_code, dtype: int64

Checking the state column

In [14]:
data.state.value_counts()

Québec              14700
QC                   6259
Quebec                160
Qu                      9
Qc                      8
-Quebec                 5
quebec                  4
Montreal, QC            2
Québec                 2
Montreal, Québec        2
魁北克省                    1
Lachine                 1
q                       1
Quebéc                  1
LaSalle,  Quebec        1
Québec City             1
魁北克                     1
québec                  1
qc                      1
canada                  1
Name: state, dtype: int64

We need to drop the columns where the state isn't listed as quebec or is one of its variant name

In [15]:
data.state.unique()

array(['QC', 'Québec', 'Quebec', 'Qc', 'Québec', '魁北克省', nan,
       'Québec City', 'Lachine', 'Montreal, Québec', 'quebec', 'canada',
       'Montreal, QC', 'LaSalle,  Quebec', 'Qu', 'québec', 'Quebéc',
       '-Quebec', 'qc', 'q', '魁北克'], dtype=object)

In [16]:
data = data[data.state.isin(['QC', 'Québec', 'Quebec', 'Qc', 'Québec', 'Québec City', 'Montreal, Québec', 'quebec', 'Montreal, QC', 'LaSalle,  Quebec', 'Qu', 'québec', 'Quebéc','-Quebec', 'qc'] )]

#df = data[data.state == ['QC', 'Québec', 'Quebec', 'Qc', 'Québec', 'Québec City', 'Montreal, Québec', 'quebec', 'Montreal, QC', 'LaSalle,  Quebec', 'Qu', 'québec', 'Quebéc','-Quebec', 'qc'] ]

In [17]:
data.state.value_counts()

Québec              14700
QC                   6259
Quebec                160
Qu                      9
Qc                      8
-Quebec                 5
quebec                  4
Montreal, QC            2
Québec                 2
Montreal, Québec        2
québec                  1
Quebéc                  1
LaSalle,  Quebec        1
Québec City             1
qc                      1
Name: state, dtype: int64

In [18]:
data.requires_license.unique()


array(['f'], dtype=object)

requires_license has all values f, its uniformative. So, we can drop that column

In [19]:
data.has_availability.value_counts()

t    21156
Name: has_availability, dtype: int64

In [20]:
data.is_business_travel_ready.value_counts()

f    21156
Name: is_business_travel_ready, dtype: int64

We can see that the above two features just carry a single value for all listings. These features wont give any new information for each listing and model cant learn anything new from them. We might as well drop them.

In [21]:
data.drop(['requires_license','has_availability','is_business_travel_ready'],axis=1, inplace=True)

In [22]:
data.to_csv('cleaned_listings_montreal.csv', header=True, index=False)
data.shape

(21156, 94)